L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,463 @@
+  // If verification fails or the peer certificate is self-signed, this will be an
+  // empty string. If verification is successful, it is a comma-separated list,
+  // where the entries are of the form "FIELD_ABBREVIATION=string"
+  // ex: "CN=testca,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU"
+  // ex: "CN=GTS Root R1,O=Google Trust Services LLC,C=US"
+  absl::string_view verified_root_cert_subject() const;
+  std::vector<absl::string_view> uri_names() const;
+  std::vector<absl::string_view> dns_names() const;
+  std::vector<absl::string_view> email_names() const;
+  std::vector<absl::string_view> ip_names() const;
+
+}
+
+// The base class of all internal verifier implementations, and the ultimate
+// class that all external verifiers will eventually be transformed into.
+// To implement a custom verifier, do not extend this class; instead,
+// implement a subclass of ExternalCertificateVerifier. Note that custom
+// verifier implementations can compose their functionality with existing
+// implementations of this interface, such as HostnameVerifier, by delegating
+// to an instance of that class.
+class CertificateVerifier {
+ public:
+  ~CertificateVerifier();
+
+  // Verifies a connection request. The check on each internal verifier could be
+  // either synchronous or asynchronous, and we will need to use return value to
+  // know.
+  //
+  // request: the verification information associated with this request
+  // callback: This will only take effect if the verifier is asynchronous.
+  //           The function that gRPC will invoke when the verifier has already
+  //           completed its asynchronous check. Callers can use this function
+  //           to perform any additional checks. The input parameter of the
+  //           std::function indicates the status of the verifier check.
+  // sync_status: This will only be useful if the verifier is synchronous.
+  //              The status of the verifier as it has already done it's
+  //              synchronous check.
+  // return: return true if executed synchronously, otherwise return false
+  bool Verify(TlsCustomVerificationCheckRequest* request,
+              std::function<void(Status)> callback,
+              Status* sync_status);
+
+  // Cancels a verification request previously started via Verify().
+  // Used when the connection attempt times out or is cancelled while an async
+  // verification request is pending.
+  //
+  // request: the verification information associated with this request
+  void Cancel(TlsCustomVerificationCheckRequest* request);
+};
+
+// A CertificateVerifier that doesn't perform any additional checks other than
+// certificate verification, if specified.
+// Note: using this solely without any other authentication mechanisms on the
+// peer identity will leave your applications to the MITM(Man-In-The-Middle)
+// attacks. Users should avoid doing so in production environments.
+class NoOpCertificateVerifier : public CertificateVerifier {
```

Comment:
This is a very "insecure" option - why are we advertising this? Can we remove this and also remove its declaration in "NoOpCertificateVerifier"?

Rationale: We should make it as hard-as-possible for users to shoot themselves in the foot.

Reply:
I'm not sure why this got implemented in the first place, let's look into the history. I'm not against removing it.
I suspect it might be a useful-for-testing construct, in which case it belongs in test code.

Reply:
Thanks, ya I strongly think we should either remove this or move it to test-only code. If a user really wants to shoot themselves in the foot, they could implement their own version of this "no-op verifier", but we shouldn't make it easy on them to do this IMO.

Reply:
Looks like this was removed.